lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,248 @@
+		e.sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	if err != nil {
+		e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// SendError sends error message to connected client.
+func (e *Engine) SendError(redisErr error) {
+	if redisErr == nil || utils.IsOKNetworkError(redisErr) {
+		return
+	}
+
+	if err := e.sendToClient(redisErr); err != nil {
+		e.Log.Errorf("failed to send message to the client: %v", err)
+		return
+	}
+}
+
+// sendToClient sends a command to connected Redis client.
+func (e *Engine) sendToClient(vals interface{}) error {
+	buf := &bytes.Buffer{}
+	wr := redis.NewWriter(buf)
+
+	if err := protocol.WriteCmd(wr, vals); err != nil {
+		return trace.BadParameter("failed to convert error to a message: %v", err)
+	}
+
+	if _, err := e.proxyConn.Write(buf.Bytes()); err != nil {
+		return trace.ConnectionProblem(err, "failed to send message to the client")
+	}
+
+	return nil
+}
+
+// HandleConnection is responsible for connecting to a Redis instance/cluster and
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	// Check that the user has access to the database.
+	err := e.authorizeConnection(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	connectionOptions, err := ParseRedisURI(sessionCtx.Database.GetURI())
+	if err != nil {
+		return trace.BadParameter("Redis connection string is incorrect: %v", err)
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	switch connectionOptions.mode {
+	case Standalone:
+		redisConn = redis.NewClient(&redis.Options{
+			Addr:      connectionAddr,
+			TLSConfig: tlsConfig,
+		})
+	case Cluster:
+		if sessionCtx.DatabaseName != "" {
+			// ref: https://redis.io/commands/select
+			return trace.BadParameter("Redis Cluster only supports database zero")
+		}
+
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	default:
+		// We've checked that while validating the config, but checking again can help with regression.
+		return trace.BadParameter("incorrect connection mode %s", connectionOptions.mode)
+	}
+
+	defer func() {
+		if err := redisConn.Close(); err != nil {
+			e.Log.Errorf("failed to close Redis connection: %v", err)
+		}
+	}()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.process(ctx, redisConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// readClientCmd reads commands from connected Redis client.
+func (e *Engine) readClientCmd(ctx context.Context) (*redis.Cmd, error) {
```

Comment:
Nit: This is minor (and probably a matter of preference) but I try to organize the functions so when you're reading the file top-to-bottom, functions are defined in the order they are called above. So I would move `readClientCmd` after `process` (in which it's called), I think it'd improve the reading experience.

Reply:
moved